test/distributed/checkpoint/e2e/test_e2e_save_and_load.py

```diff
@@ -0,0 +1,135 @@
+# Owner(s): ["oncall: distributed"]
+
+from enum import auto, Enum
+
+import torch
+import torch.distributed.checkpoint as DCP
+import torch.nn as nn
+from torch.distributed._tensor.device_mesh import init_device_mesh
+from torch.distributed.checkpoint.state_dict import get_state_dict, set_state_dict
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.api import ShardingStrategy
+from torch.distributed.tensor.parallel import PairwiseParallel, parallelize_module
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    skip_if_lt_x_gpu,
+    with_comms,
+)
+from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
+
+
+# Simple and boring model to test interface and some corner cases that do not
+# require complicated wrapping strategy.
+class TestDummyModel(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.net1 = nn.Sequential(nn.Linear(8, 16), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+        self.net3 = nn.Linear(32, 64)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+    def forward(self, x):
+        return self.net4(self.net3(self.net2(self.net1(x))))
+
+    def get_input(self):
+        return torch.rand(8, 8, device="cuda")
+
+
+class ModelType(Enum):
+    FSDP = auto()
+    HSDP = auto()
+    FSDP_TP = auto()
+
+
+class TestE2ELoadAndSave(DTensorTestBase):
+    def _create_model(self, compile, model_type):
+        dummy_model = TestDummyModel().cuda()
+
+        assert model_type in ModelType, f"{model_type} is not supported."
+        if model_type == ModelType.FSDP:
+            device_mesh = init_device_mesh(self.device_type, (self.world_size,))
+            model = FSDP(
+                dummy_model,
+                device_mesh=device_mesh,
+                use_orig_params=True,
+            )
+        elif model_type == ModelType.HSDP:
+            device_mesh = init_device_mesh(self.device_type, (2, self.world_size // 2))
+            model = FSDP(
+                dummy_model,
+                device_mesh=device_mesh,
+                use_orig_params=True,
+                sharding_strategy=ShardingStrategy.HYBRID_SHARD,
+            )
+        elif model_type == ModelType.FSDP_TP:
+            mesh_2d = init_device_mesh(
+                self.device_type, (2, self.world_size // 2), mesh_dim_names=("dp", "tp")
+            )
+            tp_mesh = mesh_2d["tp"]
+            dp_mesh = mesh_2d["dp"]
+            model = parallelize_module(dummy_model, tp_mesh, PairwiseParallel())
```

Comment:
As a follow up, we should probably switch `PairwiseParallel` to `Colwise + Rowwise` as we plan to deprecate the former soon

Reply:
@LucasLLC 

For usage of Colwise + Rowwise, you can refer to this slide. https://docs.google.com/presentation/d/1e9TNYu_u_Hz9IpfhS4_R_77de59g5AnSWEf7U-FDphg/edit#slide=id.g28cb00a468a_0_16 

So, for this model, we can specify a parallelize_plan for some of the layers, maybe layer 1 and layer 2. https://docs.google.com/presentation/d/1e9TNYu_u_Hz9IpfhS4_R_77de59g5AnSWEf7U-FDphg/edit#slide=id.g28cb00a468a_0_16

Reply:
actually since we use sequential in the module definition, I am not sure if `PairwiseParallel` is working here. So using Colwise + Rowwise might be a necessary. Basically, you need to pass in a dictionary as a plan to `parallelize_module` with its key equals to module's FQN and value is ColwiseParallel or RowwiseParallel.